Framework Cleanup (no more client code in fw)

BenDol opened the pull request:
# Description
Reworked some major cross contamination in the framework:

- Removed <client/...> includes from framework (we should never cross contaminate fw/client)

- Keep `GraphicalApplication` abstract using `ApplicationDrawEvents` & `ApplicationContext`
- Moved `ShaderManager` to the framework
- Created `UIQrCode` widget
- Discord uses update callbacks now
- Minor additions and clean up
- Replace `image-source-base64` with `image-source: base64:path/to/image`
- Added CHANGELOG.md to help with giving visibility to breaking changes
- Readd FRAMEWORK_GRAPHICS (so we can have no graphics support)
- Added creatureDiagonalWalkSpeed/playerDiagonalWalkSpeed setup config
- Added getCountOrSubType to Item lua binding
- Added startEvent to EventDispatcher (to start stored ScheduledEvent's)
- Removed g_drawPool from the application.cpp (we use dispatchPoll)
- Added startTime calculator when logging module starts
- Allow commenting out otml using #
- Added some math functions
- Added some more helper methods to UIWidget

- Added onHovered for only hovered widget events
- Make getChildIndex take child as optional parameter, if not provided it returns this widgets child index
- Added sha1Encrpyt method
- General clean up

`GraphicalApplication` was the biggest offender, we need to ensure that this is a base graphical application class and not a client specific class.

### Breaking API Changes
- `UIWidget` property `qr-code` & `qr-code-border` replaced with `UIQrCode` properties `code` & `code-border`
- `image-source-base64` replaced with `image-source: base64:/path/to/image`
- `#include "shadermanager.h"` moved to `#include <framework/graphics/shadermanager.h>`

BenDol committed:
Moved <client/...> includes from framework (we should never cross contaminate fw/client)

GraphicalApplication was the biggest offender, we need to ensure that this is a base graphical application class and not a client specific class.
Created UIQrCode widget
Minor additions and clean up

BenDol committed:
Merge remote-tracking branch 'upstream/main' into fw_cleanup

# Conflicts:
#	src/client/tile.cpp
#	src/framework/core/graphicalapplication.cpp
#	src/framework/core/graphicalapplication.h
#	src/main.cpp

BenDol committed:
Fix for latest version

BenDol committed:
Finished, no more client code in the framework!

BenDol committed:
Merge branch 'fw_update_old' into fw_cleanup

BenDol committed:
Update README

BenDol committed:
Revert

BenDol committed:
fix cmake build

BenDol committed:
Merge branch 'fw_update_old' into fw_cleanup

conde2 commented on src/framework/core/application.h:
```diff
@@ -75,6 +87,8 @@ class Application
     bool m_terminated{ false };
     bool m_stopping{ false };
+
+    ApplicationContext* m_context;
```
we can use unique_ptr to store this I belive

conde2 commented on src/framework/ui/uiwidgetimage.cpp:
```diff
@@ -36,7 +36,10 @@ void UIWidget::parseImageStyle(const OTMLNodePtr& styleNode)
+            if (node->value() == "none")
+                setImageSource("", false);
+            else
+                setImageSource(stdext::resolve_path(node->value(), node->source()), false);
```
This could be done inside setImageSource

There we could check for none, or empty string

If we do here, icons, and image-source-base64 will have different behaviour

conde2 commented on src/framework/core/graphicalapplication.h:
```diff
@@ -30,10 +30,46 @@
+
+protected:
+    uint16_t m_spriteSize;
+    ApplicationDrawEvents* m_drawEvents;
```
Here I belevie should be a shared_ptr so we don't need to delete this manually

conde2 commented on src/framework/luafunctions.cpp:
```diff
@@ -817,6 +829,14 @@ void Application::registerLuaFunctions()
     g_lua.bindClassMemberFunction<UITextEdit>("isShiftNavigation", &UITextEdit::isShiftNavigation);
     g_lua.bindClassMemberFunction<UITextEdit>("isMultiline", &UITextEdit::isMultiline);
 
+    // UITextEdit
```
Typo, should be UIQrCode

BenDol committed:
Update review suggestions

Remove `image-source-base64` in favour of `image-source: base64:/path/image`

BenDol committed:
Merge branch 'fw_update_old' into fw_cleanup

BenDol committed:
Added CHANGELOG.md to help with noting breaking changes

BenDol commented:
I have some more changes coming that are necessary in this PR, @conde2 I'll need another review when ever you get the chance, once I've committed my changes.

BenDol committed:
More framework improvements and some client improvements:

- Readd FRAMEWORK_GRAPHICS (so we can have no graphics support)
- Added creatureDiagonalWalkSpeed/playerDiagonalWalkSpeed upgrade_classification
- Added getCountOrSubType to Item lua binding
- Added startEvent to EventDispatcher (to start stored ScheduledEvent's)
- Removed g_drawPool from the application.cpp (we use dispatchPoll)
- Added startTime calculator when logging module starts
- Allow commenting out otml using #
- Added some math functions
- Added onAdopted/onAbandoned to UIWidget for when they are added to/from parent UIWidget's
- Added some more helper methods to UIWidget
- Added visibleOnly to UIWidget getChildBefore/getChildAfter methods
- Added onHovered for only hovered widget events
- Make getChildIndex take child as optional parameter, if not provided it returns this widgets child index
- Added sha1Encrpyt method
- General clean up

BenDol committed:
Rename ApplicationDrawEvents#setLoadingAsyncTexture to onLoadingAsyncTextureChanged

BenDol committed:
Update setup.otml

BenDol committed:
Added a todo

BenDol committed:
Clean up

conde2 commented on src/framework/stdext/math.cpp:
```diff
@@ -62,4 +63,39 @@ namespace stdext
+		    increment = round(v * diff);
+	    }
+	    return minNumber + increment;
+    }
```
There is a new version of this function in TFS that has fixed some issues
https://github.com/otland/forgottenserver/commit/16a985c33c80544af0ac0668bc35da11cfcaf9f9

conde2 commented on src/framework/ui/uiwidget.cpp:
```diff
@@ -247,6 +250,10 @@ void UIWidget::insertChild(size_t index, const UIWidgetPtr& child)
+    callLuaField("onInsertChild", child);
+
+    if (getRootParent() == g_ui.getRootWidget())
+        child->onAdopted(static_self_cast<UIWidget>());
```
This is going to be a spam call to lua in certain cases, I don't see much use case for this to be honest, but if there is a good reason, then it should be disabled by default like we do in tile.cpp for example:

```
    if (g_game.isTileThingLuaCallbackEnabled())
        callLuaField("onAddThing", thing);
 ```

conde2 commented on src/framework/ui/uiwidget.cpp:
```diff
@@ -606,6 +615,24 @@ void UIWidget::unlockChild(const UIWidgetPtr& child)
+        child->onAbandoned(static_self_cast<UIWidget>());
+    }
+}
+
```
Same here, its going to trigger a lot of lua calls, we should avoid this at all cost.

conde2 commented on src/framework/ui/uiwidget.cpp:
```diff
@@ -1248,22 +1303,32 @@ UIWidgetPtr UIWidget::getRootParent()
+            return child;
+    }
     return nullptr;
 }
```
The childIndex was designed to avoid looping all childrens, this breaks this behaviour.
We should create new functions for this e.g: 
getVisibleChildAfter
getVisibleChildBefore
getVisibleChildById

and to be honest I would just do it in Lua,
```
function UIWidget:getVisibleChildById(id)
  for _, child in ipairs(self:getChildren()) do
    if child:getId() == id and child:isExplicityVisible()  then
      return child
    end
  end
end
```

conde2 commented on src/framework/ui/uiwidget.cpp:
```diff
@@ -1702,6 +1768,9 @@ void UIWidget::onChildFocusChange(const UIWidgetPtr& focusedChild, const UIWidge
     callLuaField("onHoverChange", hovered);
+
+    if (hovered)
+        callLuaField("onHovered");
```
Is there any reason for this extra event? Couldn't we just check hovered in the lua side?

conde2 commented on src/framework/ui/uiwidget.cpp:
```diff
@@ -1791,6 +1860,18 @@ bool UIWidget::onDoubleClick(const Point& mousePos)
+    }
+    return nullptr;
+}
+
```
No for loop needed

you can just use:
g_ui.getHoveredWidget()

then get the id
and then getChildById

BenDol committed:
Revert old code (oops)

BenDol committed:
Reverting onAdopted, etc

BenDol committed:
Add improved version

BenDol committed:
Improve getHoveredChild

conde2 commented on src/framework/core/config.cpp:
```diff
@@ -151,6 +151,17 @@ int Config::getNodeSize(const std::string& key)
+    }
+    return node;
+}
+
```
I belive a better name for this would be getOrCreateNode

conde2 commented on src/framework/ui/uiwidget.h:
```diff
@@ -271,7 +274,7 @@ class UIWidget : public LuaObject
     bool isDisabled() { return hasState(Fw::DisabledState); }
     bool isFocused() { return hasState(Fw::FocusState); }
-    bool isHovered() { return hasState(Fw::HoverState); }
+    bool isHovered(bool orChildHovered = false) { return hasState(Fw::HoverState) || (orChildHovered && getHoveredChild() != nullptr); }
```
This could be a separated function in my opinion

BenDol committed:
Recommended fixes/clean up

BenDol committed:
Moved missed graphical sources to FRAMEWORK_GRAPHICS
